util/filter_bench.cc

```diff
@@ -0,0 +1,403 @@
+    info.reader_.reset(policy->GetFilterBitsReader(info.filter_));
+    totalMemoryUsed += info.filter_.size();
+    totalKeysAdded += keysToAdd;
+  }
+
+  uint64_t elapsedNanos = timer.ElapsedNanos();
+  double ns = (double)elapsedNanos / totalKeysAdded;
+  std::cout << "Build avg ns/key: " << ns << std::endl;
+  std::cout << "Number of filters: " << infos.size() << std::endl;
+  std::cout << "Total memory (MB): " << totalMemoryUsed / 1024.0 / 1024.0
+            << std::endl;
+
+  double bpk = totalMemoryUsed * 8.0 / totalKeysAdded;
+  std::cout << "Bits/key actual: " << bpk << std::endl;
+  if (!FLAGS_quick) {
+    double tolerableRate = std::pow(2.0, -(bpk - 1.0) / (1.4 + bpk / 50.0));
+    std::cout << "Best possible FP rate %: " << 100.0 * std::pow(2.0, -bpk)
+              << std::endl;
+    std::cout << "Tolerable FP rate %: " << 100.0 * tolerableRate << std::endl;
+
+    std::cout << "----------------------------" << std::endl;
+    std::cout << "Verifying..." << std::endl;
+
+    uint32_t outside_q_per_f = 1000000 / infos.size();
+    uint64_t fps = 0;
+    for (uint32_t i = 0; i < infos.size(); ++i) {
+      FilterInfo &info = infos[i];
+      for (uint32_t j = 0; j < info.keys_added_; ++j) {
+        always_assert(info.reader_->MayMatch(kms[0].Get(info.filter_id_, j)));
+      }
+      for (uint32_t j = 0; j < outside_q_per_f; ++j) {
+        fps +=
+            info.reader_->MayMatch(kms[0].Get(info.filter_id_, j | 0x80000000));
+      }
+    }
+    std::cout << " No FNs :)" << std::endl;
+    double prelimRate = (double)fps / outside_q_per_f / infos.size();
+    std::cout << " Prelim FP rate %: " << (100.0 * prelimRate) << std::endl;
+
+    if (!FLAGS_allow_bad_fp_rate) {
+      always_assert(prelimRate < tolerableRate);
+    }
+  }
+
+  std::cout << "----------------------------" << std::endl;
+  std::cout << "Inside queries..." << std::endl;
+  random.seed(FLAGS_seed + 1);
+  RandomQueryTest(/*inside*/ true, /*dry_run*/ true, TM_RandomFilter);
+  for (TestMode tm : testModes) {
+    random.seed(FLAGS_seed + 1);
+    RandomQueryTest(/*inside*/ true, /*dry_run*/ false, tm);
+  }
+
+  std::cout << "----------------------------" << std::endl;
+  std::cout << "Outside queries..." << std::endl;
+  random.seed(FLAGS_seed + 2);
+  RandomQueryTest(/*inside*/ false, /*dry_run*/ true, TM_RandomFilter);
+  for (TestMode tm : testModes) {
+    random.seed(FLAGS_seed + 2);
+    RandomQueryTest(/*inside*/ false, /*dry_run*/ false, tm);
+  }
+
+  std::cout << "----------------------------" << std::endl;
+  std::cout << "Done." << std::endl;
+}
+
+void FilterBench::RandomQueryTest(bool inside, bool dry_run, TestMode mode) {
+  for (auto &info : infos) {
+    info.outside_queries_ = 0;
+    info.false_positives_ = 0;
+  }
+
+  uint32_t dryRunHash = 0;
+  uint64_t maxQueries = static_cast<uint64_t>(FLAGS_m_queries) * 1000000;
+  size_t numPrimaryFilters = infos.size();
+  uint32_t primaryFilterThreshold = 0xffffffff;
```

Comment:
Can you add some comments to the function to explain how it works? Also the variables. What does "primary filter" mean is not clear when I read up to here.